### PR TITLE
fix: don't throw if unknowTargets have custom binaries

### LINKED ISSUE #4205: Prisma2 blocks prisma generate on unsupported platforms, even when providing custom binaries
This bug report is following this [conversation bit](https://github.com/prisma/prisma/issues/861#issuecomment-723810121):
> Hello there,
> 
> I'm encountering the issue described [here](https://github.com/prisma/prisma/issues/1505#issuecomment-633802109).
> 
> I understand that you are not going to officially support arm platforms yourselves and that [you just give instructions on how we can do it ourselves](https://github.com/prisma/prisma/issues/861#issuecomment-690574582).
> 
> However it seems you are blocking (probably not on purpose) any execution on non supported platforms and the workaround is to [simply add the name of the platform to a list (or 2)](https://github.com/prisma/prisma/issues/1505#issuecomment-634154657).
> 
> Could you lift that restriction when using custom binaries so we don't have to fork the entire prisma in this situation?
> 
> If needed, I can create a new issue to track this, just tell me 😃

## Bug description

Prisma2 restricts the type of platform it can be used on based on an hardcoded list, preventing the users to even try by raising an error when using `npx prisma generate`:
 ```
 Environment variables loaded from ./prisma/.env
 Warning Precompiled binaries are not available for ARM.
 Warning Precompiled binaries are not available for ARM.
 Error: Unknown binaryTargets arm-openssl-1.1.x
 ```

## How to reproduce

1. Get an unsupported platform (I'm using a raspberry pi - which is an arm platform)
2. Generate binaries for that platform following [prisma instructions](https://github.com/prisma/prisma-engines#building-prisma-engines)
3. Inject those custom binaries inside any prisma project on the said platform following [prisma instructions](https://www.prisma.io/docs/understand-prisma/under-the-hood#using-custom-engine-binaries)
4. Run `npx prisma generate`
5. See the error

## Expected behavior

[From this comment](https://github.com/prisma/prisma/issues/1505#issuecomment-634154657), it appears the blocking is purely based on a list of hardcoded predefined supported platforms.
I see 2 possible behaviours:

- Give the users a way to bypass the check on the architecture being in the list of supported ones (it will become the responsability of the user to know what he is doing) - [@janpio mentionned env var for instance](https://github.com/prisma/prisma/issues/861#issuecomment-724014072) - I don't have any insightful advice on the better method
- Bypass the check altogether as long as the user is providing his onw binaries - if this check is only useful for downloading bianries from prisma repo, and the user is already providing his own, then it is not necessary

I haven't researched into the code to know if the second option is viable.

## Prisma information

Kinda irrelevant here, but you can always have a look at [my own project](https://github.com/BaptisteLongy/shopping-list-v2)...

## Environment & setup

- OS: Raspbian GNU/Linux 10 (buster)
- Database: SQLite
- Node.js version: v14.15.0
- Prisma version:
```
Environment variables loaded from /home/pi/docker-builders/shopping-list-v2/prisma/.env
Warning Precompiled binaries are not available for arm.
@prisma/cli          : 2.10.2
@prisma/client       : 2.10.2
Current platform     : arm-openssl-1.1.x
Query Engine         : query-engine 3cc43e8ee3c92ab17f192e5874968b2b86dd6740 (at ../prisma-engines/target/release/query-engine, resolved by PRISMA_QUERY_ENGINE_BINARY)
Migration Engine     : migration-engine-cli 3cc43e8ee3c92ab17f192e5874968b2b86dd6740 (at ../prisma-engines/target/release/migration-engine, resolved by PRISMA_MIGRATION_ENGINE_BINARY)
Introspection Engine : introspection-core 3cc43e8ee3c92ab17f192e5874968b2b86dd6740 (at ../prisma-engines/target/release/introspection-engine, resolved by PRISMA_INTROSPECTION_ENGINE_BINARY)
Format Binary        : prisma-fmt 3cc43e8ee3c92ab17f192e5874968b2b86dd6740 (at ../prisma-engines/target/release/prisma-fmt, resolved by PRISMA_FMT_BINARY)
Studio               : 0.304.0
```

### PR DESCRIPTION
**Pull request by williamluke4**
Fixes #4205

### CODE REVIEW

File: src/packages/fetch-engine/src/download.ts [hunk 1]
```diff
@@ -1,27 +1,24 @@
...
 import Debug from '@prisma/debug'
-import makeDir from 'make-dir'
+import { getos, getPlatform, Platform, platforms } from '@prisma/get-platform'
+import chalk from 'chalk'
 import execa from 'execa'
+import fs from 'fs'
+import makeDir from 'make-dir'
 import pFilter from 'p-filter'
-import hasha from 'hasha'
+// Packages
```
Comment (timsuchanek) [inline src/packages/fetch-engine/src/download.ts:8]: If you reorder imports, then please put the `// Packages` to the top

Comment (williamluke4) [reply src/packages/fetch-engine/src/download.ts:8]: prettier

**Comment by habovh** [comment]
@williamluke4 It seems that Prisma still throws even when binaries are provided through `prisma/.env`:

```
$ npx prisma generate
Environment variables loaded from /home/pi/project/prisma/.env
Warning Precompiled binaries are not available for arm.
Prisma schema loaded from prisma/schema.prisma
Warning Precompiled binaries are not available for arm.
Error: Unknown binaryTarget arm-openssl-1.1.x and no custom binaries were provided
```

And here's my `prisma --version` output:

```
$ npx prisma --version
Environment variables loaded from /home/pi/project/prisma/.env
Warning Precompiled binaries are not available for arm.
@prisma/cli          : 2.12.1
@prisma/client       : 2.12.1
Current platform     : arm-openssl-1.1.x
Query Engine         : query-engine e5fe27434f5b257d9b2cb37cb622e530cbf0b624 (at ../../prisma-engines/target/release/query-engine, resolved by PRISMA_QUERY_ENGINE_BINARY)
Migration Engine     : migration-engine-cli e5fe27434f5b257d9b2cb37cb622e530cbf0b624 (at ../../prisma-engines/target/release/migration-engine, resolved by PRISMA_MIGRATION_ENGINE_BINARY)
Introspection Engine : introspection-core e5fe27434f5b257d9b2cb37cb622e530cbf0b624 (at ../../prisma-engines/target/release/introspection-engine, resolved by PRISMA_INTROSPECTION_ENGINE_BINARY)
Format Binary        : prisma-fmt e5fe27434f5b257d9b2cb37cb622e530cbf0b624 (at ../../prisma-engines/target/release/prisma-fmt, resolved by PRISMA_FMT_BINARY)
Studio               : 0.322.0
```

Am I missing something here? Another way to provide custom binaries maybe?

**Comment by williamluke4** [comment]
Interesting, Let me take a look

**Comment by williamluke4** [comment]
@habovh Could you have a look into `binaryNeedsToBeDownloaded` and see where it is returning true?

**Comment by habovh** [comment]
@williamluke4 it seems that the .env is simply ignored by Prisma at this point. Here's what I get after a bit of logging:

```
$ npx prisma generate
Environment variables loaded from /home/pi/project/prisma/.env
Warning Precompiled binaries are not available for arm.
Prisma schema loaded from prisma/schema.prisma
Warning Precompiled binaries are not available for arm.
LOG: await exists() with job.targetFilePath= /home/jabe/graph-api/node_modules/@prisma/cli/query-engine-arm-openssl-1.1.x
LOG: targetExists: false
LOG: !targetExists, returning true
Error: Unknown binaryTarget arm-openssl-1.1.x and no custom binaries were provided
error Command failed with exit code 1.
```

As you can see from my `LOG` prefixed output, the `job.targetFilePath` is not the one that `npx prisma --version` outputs. It seems that Prisma ignores the environment variables here.

**Comment by williamluke4** [comment]
This should fix it https://github.com/prisma/prisma/pull/4413

**Pull request merged**